src/ReduxCanvasV4.js

```diff
@@ -0,0 +1,61 @@
+import Immutable from 'immutable'
+import React from 'react'
+import { Provider, connect } from 'react-redux'
+import createStore from './createStore'
+import Pixel from './Pixel'
+
+// Reducer
+const store = createStore((state = Immutable.Map(), action) => {
+  if (action.type === 'TOGGLE') {
+    const key = action.i + ',' + action.j
+    return state.set(key, !state.get(key))
+  }
+  return state
+})
+
+const ACTIVE_PROPS = { active: true }
+const INACTIVE_PROPS = { active: false }
+
+// Connected pixel
+const ConnectedPixel = connect(
+  (initialState, initialProps) => {
+    const { i, j } = initialProps
+    return state => {
```

Comment:
We know `i` and `j` never change so we use a fast path: return a selector that doesn't look at its prop. Selectors that look at `ownProps` are much slower.

Reply:
Hi @gaearon, not really getting this change, you're still using `i` and `j` here, how is this not looking at `ownProps`?

Reply:
Because it pulls `i` and `j` out of `ownProps` from the _initial_ call, but then returns a function that only uses a single `(state)` argument.  That way, `connect` will only run the returned function when the store state has actually changed, and not when the component's props have changed.

Reply:
@markerikson Surely if props have changed then the whole `mapStateToProps` would be re-run returning a new instance of that function?

Is the fact that the function returns always the same `ACTIVE_PROPS` or `INACTIVE_PROPS` of any significance in all of that?

Reply:
No, that's the point of the "factory function" syntax.

If the first call to `mapState` returns a function instead of an object, then `connect` will use the returned function instead of the original function.

In either case, if the final `mapState` function has a signature of `(state, ownProps)`, it will be called whenever either the store state or the wrapper's props have changed.  If it has a signature of `(state)`, it will _only_ be called when the store state has changed.

So, in this example, the _first_ `mapState` function has a `(state, ownProps)` signature, but the _returned_ function only has a `(state)` signature, so it will be called less often.

As far as actually re-rendering, `connect` just checks to see if the object returned from `mapState` is shallow-equal to the previous result.  Defining these results as constant references doesn't really matter, other than avoiding an object allocation each time.

Please see [the React-Redux docs on using `mapState` functions](https://react-redux.js.org/using-react-redux/connect-mapstate) for more details, as well as my blog post [Idiomatic Redux: The Implementation and History of React-Redux](https://blog.isquaredsoftware.com/2018/11/react-redux-history-implementation/).

Reply:
@markerikson Many thanks for quick response. I was confused thinking that the optimisation is done by returning `ACTIVE_PROPS` and `INACTIVE_PROPS` instead of the object, whereas the real optimisation is turned on by using a `mapStateToProps` function taking one argument instead of two.

I understand the same optimization would apply if `mapStateToProps` returned an arrow function taking one argument instead of two, as was the case originally?

You are right, it's all written in the documentation but I read it so many times yet failed to understand how the optimization works. I guess it's because it's the first time I am seeing some functionality/optimization enabled by the fact that a callback takes one instead of two arguments.

Reply:
Yep, that's the case.